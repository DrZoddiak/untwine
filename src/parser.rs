use std::marker::PhantomData;

use crate::{context::ParserContext, result::ParserResult};

pub trait Parser<'p, C: 'p, T: 'p, E: 'p>: private::SealedParser<C, T, E> {
    fn parse(&self, ctx: &'p ParserContext<'p, C>) -> ParserResult<T, E>;

    fn map<V: 'p>(self, f: impl Fn(T) -> V + 'p) -> impl Parser<'p, C, V, E> + 'p
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| self.parse(ctx).map(&f))
    }

    fn optional(self) -> impl Parser<'p, C, Option<T>, E>
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| {
            let res = self.parse(ctx);
            ParserResult::new(Some(res.success), res.error, res.pos)
        })
    }

    fn or(self, other: impl Parser<'p, C, T, E> + 'p + Sized) -> impl Parser<'p, C, T, E>
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| {
            let res = self.parse(ctx);
            if res.success.is_some() {
                return res;
            }
            other.parse(ctx).integrate_error(res)
        })
    }

    fn repeating(self) -> impl Parser<'p, C, Vec<T>, E>
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| {
            let mut elems = vec![];
            let res = self.parse(ctx);
            let Some(elem) = res.success else {
                return ParserResult::new(None, res.error, res.pos);
            };
            elems.push(elem);

            let failed = loop {
                let res = self.parse(ctx);
                let Some(elem) = res.success else {
                    break res;
                };
                elems.push(elem);
            };
            ParserResult::new(Some(elems), failed.error, failed.pos)
        })
    }

    fn optional_repeating(self) -> impl Parser<'p, C, Vec<T>, E>
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| {
            let mut elems = vec![];
            let failed = loop {
                let res = self.parse(ctx);
                let Some(elem) = res.success else {
                    break res;
                };
                elems.push(elem);
            };
            ParserResult::new(Some(elems), failed.error, failed.pos)
        })
    }

    fn delimited<D>(self, delim: impl Parser<'p, C, D, E> + 'p) -> impl Parser<'p, C, Vec<T>, E>
    where
        Self: Sized + 'p,
        D: 'p,
    {
        parser(move |ctx| {
            let mut elems = vec![];
            let mut res = self.parse(ctx);

            let Some(elem) = res.success.take() else {
                return ParserResult::new(Some(elems), res.error, res.pos);
            };
            elems.push(elem);

            let mut last_res = res;
            while delim.parse(ctx).success.is_some() {
                let mut res = self.parse(ctx);
                let Some(elem) = res.success.take() else {
                    return ParserResult::new(None, res.error, res.pos).integrate_error(last_res);
                };
                last_res = res;
                elems.push(elem);
            }
            ParserResult::new(Some(elems), last_res.error, last_res.pos)
        })
    }

    fn optional_delimited<D>(
        self,
        delim: impl Parser<'p, C, D, E> + 'p,
    ) -> impl Parser<'p, C, Vec<T>, E>
    where
        Self: Sized + 'p,
        D: 'p,
    {
        parser(move |ctx| {
            let mut elems = vec![];

            let mut res = self.parse(ctx);
            let Some(elem) = res.success.take() else {
                return ParserResult::new(Some(elems), res.error, res.pos);
            };
            elems.push(elem);

            let mut last_res = res;
            while delim.parse(ctx).success.is_some() {
                let mut res = self.parse(ctx);
                let Some(elem) = res.success.take() else {
                    return ParserResult::new(None, res.error, res.pos).integrate_error(last_res);
                };
                last_res = res;
                elems.push(elem);
            }
            ParserResult::new(Some(elems), last_res.error, last_res.pos)
        })
    }

    fn ignore(self) -> impl Parser<'p, C, (), E>
    where
        Self: Sized + 'p,
    {
        self.map(|_| ())
    }

    fn ignore_err(self) -> impl Parser<'p, C, T, E>
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| {
            let mut res = self.parse(ctx);
            res.error = None;
            res
        })
    }

    fn span(self) -> impl Parser<'p, C, &'p str, E>
    where
        Self: Sized + 'p,
    {
        parser(move |ctx| {
            let start = ctx.cursor();
            let res = self.parse(ctx);
            res.map(|_| &ctx.input[start..ctx.cursor()])
        })
    }

    fn unilateral(self) -> impl Parser<'p, C, T, E>
    where
        Self: Sized + 'p,
        E: std::fmt::Debug,
    {
        parser(move |ctx| {
            let start = ctx.cursor();
            let res = self.parse(ctx);
            if res.success.is_none() {
                ctx.reset(start);
            }
            res
        })
    }
}

struct ParserImpl<'p, F, C, T, E>(F, PhantomData<&'p (C, T, E)>)
where
    F: Fn(&'p ParserContext<'p, C>) -> ParserResult<T, E> + 'p,
    T: 'p,
    E: 'p;

pub fn parser<'p, C, T, E>(
    f: impl Fn(&'p ParserContext<'p, C>) -> ParserResult<T, E> + 'p,
) -> impl Parser<'p, C, T, E> + 'p
where
    C: 'p,
    T: 'p,
    E: 'p,
{
    ParserImpl(f, PhantomData)
}

mod private {
    pub trait SealedParser<C, T, E> {}
}

impl<'p, C, T, E, P> private::SealedParser<C, T, E> for P
where
    C: 'p,
    E: 'p,
    P: Parser<'p, C, T, E>,
    T: 'p,
{
}

impl<'p, F, C, T, E> Parser<'p, C, T, E> for ParserImpl<'p, F, C, T, E>
where
    F: Fn(&'p ParserContext<'p, C>) -> ParserResult<T, E>,
    T: 'p,
    E: 'p,
{
    fn parse(&self, ctx: &'p ParserContext<'p, C>) -> ParserResult<T, E> {
        (self.0)(ctx)
    }
}
