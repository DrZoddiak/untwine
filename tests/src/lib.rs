#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use untwine::parser;

    parser! {
        num: num=<'0'-'9'+> -> u32 { num.parse().unwrap() }
        pub num_list = num$","+ -> Vec<u32>;
    }

    fn operate(left: f64, op: char, right: f64) -> f64 {
        match op {
            '+' => left + right,
            '-' => left - right,
            '/' => left / right,
            '*' => left * right,
            _ => unreachable!(),
        }
    }

    parser! {
        sep = #{char::is_ascii_whitespace}*;
        num: num=<"-"? '0'-'9'+ ("." '0'-'9'+)?> -> f64 { num.parse().unwrap() }
        term = (num | "(" sep expr sep ")") -> f64;
        add: first=mul sep ops=(["+-"] sep mul)* -> f64 { ops.into_iter().fold(first, |left, (op, right)| operate(left, op, right)) }
        mul: first=term sep ops=(["*/"] sep term)* -> f64 { ops.into_iter().fold(first, |left, (op, right)| operate(left, op, right)) }
        pub expr = add -> f64;
    }

    #[test]
    pub fn test_num_list() {
        assert_eq!(untwine::parse(num_list, "1,2,3").unwrap(), vec![1, 2, 3]);
        assert_snapshot!(untwine::parse_pretty(num_list, "1,2,", false).unwrap_err());
    }

    #[test]
    pub fn test_expr() {
        assert_eq!(untwine::parse(expr, "1+2*3").unwrap(), 7.0);
        assert_eq!(untwine::parse(expr, "1+(3/4)").unwrap(), 1.75);
        assert_eq!(untwine::parse(expr, "4--1").unwrap(), 5.0);
        assert_snapshot!(untwine::parse_pretty(expr, "1.", false).unwrap_err());
        assert_snapshot!(untwine::parse_pretty(expr, "(1\n\n+", false).unwrap_err());
        assert_snapshot!(untwine::parse_pretty(expr, "(1\n\n+5", false).unwrap_err());
        assert_snapshot!(untwine::parse_pretty(expr, "--1", false).unwrap_err());
    }
}
