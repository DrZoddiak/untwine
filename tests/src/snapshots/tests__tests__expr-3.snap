---
source: tests/src/lib.rs
expression: "untwine::parse_pretty(expr, \"(1\\n\\n+5\", false).unwrap_err()"
---
    | beginning here
    v
1 | (1
  | ...
3 | +5
    --^
[3:2] Expected literal ')'
